File: backend-spring/src/main/java/com/patchflow/controller/NotificationController.java
package com.patchflow.controller;

import com.patchflow.config.Auth;
import com.patchflow.entity.Notification;
import com.patchflow.entity.User;
import com.patchflow.service.NotificationService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.Map;

@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
public class NotificationController {

    private final NotificationService notificationService;

    @GetMapping
    public ResponseEntity<?> getMyNotifications(HttpServletRequest req) {
        User user = Auth.require(req);
        return ResponseEntity.ok(notificationService.getUserNotifications(user.getUserId()));
    }

    @PatchMapping("/{id}/read")
    public ResponseEntity<?> markAsRead(@PathVariable String id, HttpServletRequest req) {
        User user = Auth.require(req);
        try {
            Notification n = notificationService.markAsRead(id, user.getUserId());
            return ResponseEntity.ok(n);
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }
}

File: backend/src/controllers/notification.controller.ts
import { Response } from "express";
import { NotificationService } from "../services/notification.service";
import { AuthenticatedRequest } from "../middlewares/auth.middleware";

export class NotificationController {
  static async getMyNotifications(req: AuthenticatedRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).json({ error: "Unauthorized" });

      const notifications = await NotificationService.getUserNotifications(req.user.userId);
      return res.status(200).json(notifications);
    } catch (error) {
      console.error(error);
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async markAsRead(req: AuthenticatedRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).json({ error: "Unauthorized" });

      const notificationId = req.params.id as string;
      const updated = await NotificationService.markAsRead(notificationId, req.user.userId);
      
      return res.status(200).json(updated);
    } catch (error: any) {
      if (error.message === "Notification not found") {
        return res.status(404).json({ error: error.message });
      }
      return res.status(500).json({ error: "Internal server error" });
    }
  }
}
